src/lib.rs

```diff
@@ -80,32 +80,53 @@ mod panic;
 ///
 /// It's not required to call this before the library is used, but it's
 /// recommended to do so as soon as the program starts.
+#[inline]
 pub fn init() {
+    /// An exported constructor function. On supported platforms, this will be
+    /// invoked automatically before the program's `main` is called.
+    #[cfg_attr(any(target_os = "linux", target_os = "freebsd", target_os = "android"), link_section = ".init_array")]
+    #[cfg_attr(target_os = "macos", link_section = "__DATA,__mod_init_func")]
+    #[cfg_attr(target_os = "windows", link_section = ".CRT$XCU")]
+    static INIT_CTOR: extern "C" fn() = init_inner;
+
+    /// Used to prevent concurrent calls when initializing after `main` starts.
     static INIT: Once = Once::new();
-    INIT.call_once(|| {
-        platform_init();
-        unsafe {
-            assert_eq!(curl_sys::curl_global_init(curl_sys::CURL_GLOBAL_ALL), 0);
-        }
 
-        // Note that we explicitly don't schedule a call to
-        // `curl_global_cleanup`. The documentation for that function says
-        //
-        // > You must not call it when any other thread in the program (i.e. a
-        // > thread sharing the same memory) is running. This doesn't just mean
-        // > no other thread that is using libcurl.
+    // We invoke our init function through our static to ensure the symbol isn't
+    // optimized away by a bug: https://github.com/rust-lang/rust/issues/47384
+    INIT.call_once(|| INIT_CTOR());
```

Comment:
Perhaps this `INIT `could move into `init_inner`? That way we wouldn't need a second atomic I think?

Reply:
I don't think so, because we are not allowed to call `Once::call_once` inside a constructor function, because `Once::call_once` depends on the Rust runtime. But the constructor function will run before the Rust runtime is initialized.

Reply:
I could test it on the various platforms, but `Once` relies on the `std::thread` machinery to be available.

An alternative would be to replace `Once` with an equivalent that does not rely on the standard library. Then we could use the same concurrency primitive in both the constructor and in the lazy init call.

Reply:
AFAIK there's no real "runtime" requirement and by definition this is happening on just the main thread so there's no contention, so it won't be needed unless this is manually called after the main function?

Reply:
> AFAIK there's no real "runtime" requirement

Basically, we have to _know_ that `Once` will work correctly when used before these calls are made: https://github.com/rust-lang/rust/blob/master/src/libstd/rt.rs#L34-L45

> and by definition this is happening on just the main thread so there's no contention, so it won't be needed unless this is manually called after the main function?

Right, but we still do the lazy init check in `Easy` and friends, so we have to make sure that the lazy init calls don't initialize a second time with something like a `Once` or an atomic. The way the code is right now, the constructor functions will set a boolean atomic if they ran since they can't inform the `Once` of that fact.

Reply:
OK, based on my reading of the [`Once` source](https://github.com/rust-lang/rust/blob/master/src/libstd/sync/once.rs), it should be OK to use since there's no possibility for contention. If there were, it would call `thread::current()` which to me looks like it would not work as expected if invoked before `rt.rs`. I'm just not sure if its OK to rely on the `Once` implementation details like that.

Reply:
Even if `thread::current()` is called, however, it shoudl be valid to do so. That's lazily initialized if it's not already present.

Reply:
Maybe this is getting in the weeds a bit, but wouldn't that cause this assert to fail? https://github.com/rust-lang/rust/blob/master/src/libstd/sys_common/thread_info.rs#L40 Since `lang_start_internal` calls `thread_info::set`, but the `ThreadInfo` for the main thread would have been already lazily created by the prior call to `thread::current()`?

Reply:
The following program:

```rust
#[used]
#[cfg_attr(target_os = "macos", link_section = "__DATA,__mod_init_func")]
static CTOR: extern "C" fn() = init;

extern "C" fn init() {
    // Yay side effects
    std::thread::current();
}

fn main() {
    println!("main");
}
```

unconditionally panics on start on macOS with the following output:

```
thread '<unnamed>' panicked at 'assertion failed: c.borrow().is_none()', src/libstd/sys_common/thread_info.rs:40:26
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
fatal runtime error: failed to initiate panic, error 5
fish: 'cargo run' terminated by signal SIGABRT (Abort)
```

Reply:
Hm I would classify that as a bug in the standard library actually. In any case that seems fine to avoid it for now while we wait for that to get fixed. Want to fixup CI here and then we can merge?

Reply:
Yeah, the CI failures are from ctest being broken on nightly now because the `asm!` macro was removed. I can look into that but I'm not sure where to begin.

Reply:
Ah I was also seeing rustfmt issues. The issue looks like it's the `extprim` crate, which I don't really have time to dig into where it's coming from right now.

I think ctest is largely unmaintained by gnzlbg now. I started a fork of it at https://github.com/alexcrichton/ctest/tree/use-maybe-unini and we can add more patches there to use here if necessary.

Reply:
Yeah I just opened a separate issue for the extprim problem: #337